### PR TITLE
feat(VoteProgram): support VoteInstruction::Authorize

### PR DESCRIPTION
**Pull request by mooori**
#### Problem

Solana-web3.js does not support [`VoteInstruction::Authorize`](https://docs.rs/solana-vote-program/1.9.5/solana_vote_program/vote_instruction/enum.VoteInstruction.html#variant.Authorize).

#### Summary of Changes
Support it by adding a constructor and decoder for `VoteInstruction::Authorize`.

Fixes feature request in [#22932 (comment)](https://github.com/solana-labs/solana/pull/22932#pullrequestreview-873547038).

### CODE REVIEW

File: web3.js/test/vote-program.test.ts [hunk 1]
```diff
@@ -154,13 +173,57 @@ describe('VoteProgram', () => {
...
+        votePubkey: newVoteAccount.publicKey,
+        authorizedPubkey: authorized.publicKey,
+        newAuthorizedPubkey: newAuthorized.publicKey,
+        voteAuthorizationType: VoteAuthorizationLayout.Withdrawer,
+      });
+      await sendAndConfirmTransaction(connection, authorize, [authorized], {
+        preflightCommitment: 'confirmed',
+      });
+
+      // Test old authorized cannot witdraw anymore.
```
Comment (CriesofCarrots) [inline web3.js/test/vote-program.test.ts:215]:
```suggestion
      // Test old authorized cannot withdraw anymore.
```

File: web3.js/test/vote-program.test.ts [hunk 2]
```diff
@@ -154,13 +173,57 @@ describe('VoteProgram', () => {
...
+        votePubkey: newVoteAccount.publicKey,
+        authorizedWithdrawerPubkey: authorized.publicKey,
+        lamports: minimumAmount,
+        toPubkey: recipient.publicKey,
+      });
+      await expect(
+        sendAndConfirmTransaction(connection, withdraw, [authorized], {
+          preflightCommitment: 'confirmed',
+        }),
+      ).to.be.rejected;
```
Comment (CriesofCarrots) [inline web3.js/test/vote-program.test.ts:226]: Do you mind adding one more transaction to demonstrate a successful withdraw with the new authority?

File: web3.js/test/vote-program.test.ts [hunk 3]
```diff
@@ -154,13 +173,57 @@ describe('VoteProgram', () => {
...
       });
-
       await sendAndConfirmTransaction(connection, withdraw, [authorized], {
         preflightCommitment: 'confirmed',
       });
       expect(await connection.getBalance(recipient.publicKey)).to.eq(
         LAMPORTS_PER_SOL,
       );
+
+      const newAuthorized = Keypair.generate();
```
Comment (CriesofCarrots) [inline web3.js/test/vote-program.test.ts:183]: Can you also add a case to test that the authorized withdrawer can sign to authorize a new Voter? More of a demo to the user than testing anything in your implementation.
ref: https://github.com/solana-labs/solana/pull/22522
Might need to set up separate newVoter and newWithdrawer Keypairs here.

**Review by CriesofCarrots — COMMENTED** [review]
Impl looks great! One typo nit, plus a couple cases for the integration test that I think would be helpful, if you're willing?

**Comment by mooori** [comment]
I've extended the integration test and fixed the typo.

When trying to authorize a new Voter more than once per epoch, the transaction fails with [`VoteError::TooSoonToReauthorize`](https://docs.rs/solana-vote-program/1.9.5/solana_vote_program/vote_instruction/enum.VoteError.html#variant.TooSoonToReauthorize).

To let Witdrawer authorize a new Voter, I removed the test of Voter authorizing a new Voter (as it's quite straighforward). Alternatively, to keep that test, a new Vote account could be created.

**Comment by mooori** [comment]
I think the CI failure (except web3-commit-lint) is not related to this PR and locally `test:live-with-test-validator` passes.

@CriesofCarrots can you see if there's something wrong on my side or if it's an issue with CI?

**Comment by CriesofCarrots** [comment]
> When trying to authorize a new Voter more than once per epoch, the transaction fails with [`VoteError::TooSoonToReauthorize`](https://docs.rs/solana-vote-program/1.9.5/solana_vote_program/vote_instruction/enum.VoteError.html#variant.TooSoonToReauthorize).

Ah yeah, true. Probably not worth mucking with the epoch or adding a second test for that. What you've done seems fine to me, thanks!

> can you see if there's something wrong on my side or if it's an issue with CI?

At a glance, it doesn't look like anything to do with your changes. I'll take a closer look and get back to you. May ask you to rebase, but otherwise this seems gtg.

**Comment by CriesofCarrots** [comment]
@mooori , should be fixed now. Can you rebase and push plz?

**Comment by mooori** [comment]
> @mooori , should be fixed now. Can you rebase and push plz?

Done and now CI passed :)

Thank you!

**Review by CriesofCarrots — APPROVED** [review]
Thank you!

**Pull request merged**